### PR TITLE
Improve std::hash docs

### LINKED ISSUE #29357: API Docs: hash
Part of https://github.com/rust-lang/rust/issues/29329

http://doc.rust-lang.org/std/hash/

Here's what needs to be done to close out this issue: 

- [ ] [`BuildHasherDefault`](https://doc.rust-lang.org/std/hash/struct.BuildHasherDefault.html) needs a stronger summary/explanation distinction. "This structure is zero-sized and does not need construction." is weird.
- [ ] [`BuildHasher`](https://doc.rust-lang.org/std/hash/trait.BuildHasher.html) also needs a stronger summary/explanation distinction.  It also needs a better explanation of why it exists with an example.
- [ ] [`Hash`](https://doc.rust-lang.org/std/hash/trait.Hash.html) needs general improvement, it is the main bit of this module, but its docs feel tossed together and have few examples or coherence.
- [ ] [`Hasher`](https://doc.rust-lang.org/std/hash/trait.Hasher.html) needs more than just a summary line; it needs explanation and examples.

### PR DESCRIPTION
**Pull request by chordowl**
Fixes #29357.

For details on what exactly I've done, see the commit descriptions.

There are some things I'm not sure about, but would like to address before merging this so the issue can be closed; any feedback on these points would really be appriciated:
* [x] ~I didn't touch the module level docs at all. On the one hand, I think they could use a short overview over the module; on the other hand, the module really isn't that big and I don't know if I could really do anything beyond just duplicating the type's summaries...~
* [x] ~I feel like the module-level examples are quite long-winded and not to the point, but I couldn't really think of anything better. Any ideas?~
* [x] ~Should `Hasher` get an example for implementing it? There is one in the module documentation, but it only "implements" it via `unimplemented!` and I'm not sure what the value of that is.~
* [x] ~Should `Hasher`'s `write_{int}` methods get examples?~

If there's anything else you'd like to see in std::hash's docs, please let me know!

r? @rust-lang/docs

### CODE REVIEW

File: src/libcore/hash/mod.rs [hunk 1]
```diff
@@ -258,12 +352,35 @@ pub trait Hasher {
...
-/// Note that for each instance of `BuildHasher`, the created hashers should be
-/// identical. That is, if the same stream of bytes is fed into each hasher, the
-/// same output will also be generated.
+/// hasher_1.write_u32(8128);
+/// hasher_2.write_u32(8128);
+///
+/// assert_eq!(hasher_1.finish(), hasher_2.finish());
+/// ```
+///
+/// [`build_hasher`]: #method.build_hasher
```
Comment (lukaramu) [inline src/libcore/hash/mod.rs]: This needs to link to `#tymethod.build_hasher` (I forgot to run linkchecker locally >_>)

File: src/libcore/hash/mod.rs [hunk 2]
```diff
@@ -107,29 +107,25 @@ mod sip;
...
-/// This trait can be used with `#[derive]` if all fields implement `Hash`.
-/// When `derive`d, the resulting hash will be the combination of the values
-/// from calling [`.hash`] on each field.
-///
-/// ## How can I implement `Hash`?
+/// ```
+/// #[derive(Hash)]
+/// struct Rustacean {
+///    name: String,
+///    country: String,
```
Comment (frewsxcv) [inline src/libcore/hash/mod.rs]: We tend to use four spaces for indentation.

File: src/libcore/hash/mod.rs [hunk 3]
```diff
@@ -107,29 +107,25 @@ mod sip;
...
+/// }
+/// ```
 ///
-/// If you need more control over how a value is hashed, you need to implement
-/// the `Hash` trait:
+/// If you need more control over how a value is hash, you can of course
```
Comment (frewsxcv) [inline src/libcore/hash/mod.rs]: "how a value is **hashed**"

**Comment by lukaramu** [comment]
I have addressed the broken link as well as the indent and grammar slipups that were brought up.

Since I personally have barely any idea on how to act on the issues I outline in the initial comment, and as they weren't on the list on the issue anyway, I've decided not to address them within this PR (unless that is a dealbreaker for someone :sweat_smile:).

**Comment by frewsxcv** [comment]
This is great @lukaramu, thanks so much for this! 🎉 

@bors r+ rollup

**Comment by frewsxcv** [comment]
@bors retry

prioritizing rollup

**Pull request merged**